src/Umbraco.Web.Website/Controllers/UmbLoginController.cs

```diff
@@ -33,12 +51,133 @@ public class UmbLoginController : SurfaceController
+        IUmbracoContextAccessor umbracoContextAccessor,
+        IUmbracoDatabaseFactory databaseFactory,
+        ServiceContext services,
+        AppCaches appCaches,
+        IProfilingLogger profilingLogger,
+        IPublishedUrlProvider publishedUrlProvider,
+        IMemberSignInManager signInManager)
+        : this(
+            umbracoContextAccessor,
+            databaseFactory,
+            services,
+            appCaches,
+            profilingLogger,
+            publishedUrlProvider,
+            signInManager,
+            StaticServiceProvider.Instance.GetRequiredService<IMemberManager>(),
+            StaticServiceProvider.Instance.GetRequiredService<ITwoFactorLoginService>(),
+            StaticServiceProvider.Instance.GetRequiredService<IEmailSender>(),
+            StaticServiceProvider.Instance.GetRequiredService<IHttpContextAccessor>(),
+            StaticServiceProvider.Instance.GetRequiredService<LinkGenerator>(),
+            StaticServiceProvider.Instance.GetRequiredService<IOptions<WebRoutingSettings>>(),
+            StaticServiceProvider.Instance.GetRequiredService<IOptions<GlobalSettings>>())
+    {
+    }
+
+    [HttpPost]
+    [ValidateAntiForgeryToken]
+    [ValidateUmbracoFormRouteString]
+    public async Task<IActionResult> HandleForgottenPassword([Bind(Prefix = "forgottenPasswordModel")] ForgottenPasswordModel model)
+    {
+        if (ModelState.IsValid == false)
+        {
+            return CurrentUmbracoPage();
+        }
+
+        MergeRouteValuesToModel(model);
+
+        if (!_emailSender.CanSendRequiredEmail())
+        {
+            return CurrentUmbracoPage();
+        }
+
+        MemberIdentityUser? memberIdentity = await _memberManager.FindByNameAsync(model.Username);
+        if (memberIdentity == null!)
+        {
+            return new ValidationErrorResult(
+                $"No local member found for username {model.Username}");
+        }
+
+        // send the email
+        await SendMemberResetEmailAsync(memberIdentity, null, null, null);
+
+        TempData["ForgottenPasswordSuccess"] = true;
+
+        return CurrentUmbracoPage();
+    }
+
+    [HttpPost]
+    [ValidateAntiForgeryToken]
+    [ValidateUmbracoFormRouteString]
+    public async Task<IActionResult> HandleResetPassword([Bind(Prefix = "resetPasswordModel")] ResetPasswordModel model)
+    {
+        if (ModelState.IsValid == false)
+        {
+            return CurrentUmbracoPage();
+        }
+
+        MergeRouteValuesToModel(model);
+
+        string? token = model.Token.FromUrlBase64();
+
+        if (string.IsNullOrWhiteSpace(token) || string.IsNullOrWhiteSpace(model.UserId))
+        {
+            return CurrentUmbracoPage();
+        }
+
+        MemberIdentityUser? memberIdentity = await _memberManager.FindByIdAsync(model.UserId);
+        if (memberIdentity == null!)
+        {
+            return new ValidationErrorResult(
+                $"No local member found for user id {model.UserId}");
```

Comment:
Similar to requesting a password reset, shouldn't reveal that the specific Id does not exist.